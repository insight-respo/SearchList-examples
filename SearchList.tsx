/**
 * @fileoverview 搜索列表的高阶组件
 * @author 蒋璇 <jiangxuan@dxy.cn>
 */

// tslint:disable: member-ordering
import { TABLE_DEFAULT_PAGE_SIZE } from '@/constants';
import { FormComponentProps } from 'antd/lib/form';
import { PaginationProps } from 'antd/lib/pagination';
import { SorterResult } from 'antd/lib/table';
import { Location } from 'history';
import { isMoment } from 'moment';
import * as React from 'react';
import Router from 'umi/router';

/**
 * @param T list 的一条数据的 结构
 */
export interface ISearchListOptions<T> {
  /**
   * @description 单选项的字段, 场景
   * * 是否关联成分词, 是否关联靶点... antd 的 table 在单选的场景下依旧是数组, 目前 insight 接口规范中需要是
   * 一个值, 而不是数组, 详情见规范 https://jiang-xuan.github.io/jfrontlife/前后端交互规范#请求规范
   */
  singleFields?: Array<keyof T>;
  /**
   * @description 在 是否关联成分词, 是否关联靶点 等的场景下, 一般不会直接已字段名进行请求, 而是会加上一个 `isAssoc` 的前缀
   * 该配置可以用来配置该前缀
   */
  singleFieldPrefix?: string;
  /**
   * @description 在筛选中中有时会存在 时间段 这样的筛选项, 目前 insight 接口规范中对于时间段的参
   * 数会拆分, 而不是作为一个数组, 详情见规范: https://jiang-xuan.github.io/jfrontlife/前后端交互规范#请求规范
   */
  timeRangeFields?: Array<keyof T>;
  /**
   * @description 请求列表数据的 dispatch 的 type
   */
  listRequestType: string;
}

export interface ISearchListProps extends FormComponentProps {
  dispatch: (options: { type: string; payload: object }) => void;
  location: Location;
}

/**
 * @description 用于给调用方进行 type 断言
 */
export interface ISearchListWrappedComponentProps<T> {
  handleSubmit(event: React.SyntheticEvent): void;
  handleTableChange(
    pagination: PaginationProps,
    filters: Record<keyof T, string[]>,
    sorter: SorterResult<T>
  ): void;
  handleSearch(): void;
  urlQueryParams: Record<keyof T, string> & { current: string, pageSize: string };
}

export default function<T>(options: ISearchListOptions<T>) {
  const { singleFields, singleFieldPrefix, timeRangeFields, listRequestType } = options;
  return WrappedComponent => {
    const wrappedComponentName =
      WrappedComponent.displayName || WrappedComponent.name || 'Component';
    // 实际组件
    const component = class SearchListHOC extends React.Component<ISearchListProps> {
      public static displayName: string = `SearchList(${wrappedComponentName})`;

      private tablePagination: PaginationProps;

      private tableFilters: Record<keyof T, string[]>;

      private tableSorter: SorterResult<T>;

      private formValues: object;

      private initUrlQueryParams = () => {
        const {
          location: { query },
        } = this.props;

        return query as object;
      };

      private urlQueryParams = this.initUrlQueryParams();

      private handleTableChange = (
        pagination: PaginationProps,
        filters: Record<keyof T, string[]>,
        sorter: SorterResult<T>
      ) => {
        this.tablePagination = pagination;
        this.tableFilters = filters;

        if (sorter.field) {
          this.tableSorter = sorter;
        }

        this.handleSearch();
      };

      private handleSubmit = (event: React.SyntheticEvent) => {
        event.preventDefault();

        this.props.form.validateFields((errors, values) => {
          if (errors) {
            return;
          }

          this.formValues = values;
          this.handleSearch();
        });
      };

      private handleSearch = () => {
        const {
          dispatch,
          location: { pathname },
        } = this.props;

        let params = { ...this.urlQueryParams };

        if (this.formValues) {
          const formatedFormValue = Object.entries(this.formValues).reduce((prev, curr) => {
            const [key, value] = curr;
            // 如果数据是一个数组, 那么就说明该数据是一个时间 range 选择器
            prev[key] = Array.isArray(value)
              ? value.map(item => (isMoment(item) ? item.valueOf() : item))
              : value;

            return prev;
          }, {});
          params = { ...params, ...formatedFormValue };
        }

        if (this.tablePagination) {
          params = {
            ...params,
            current: this.tablePagination.current,
            pageSize: this.tablePagination.pageSize,
          };
        } else {
          params = {
            ...params,
            current: 1,
            pageSize: TABLE_DEFAULT_PAGE_SIZE,
          };
        }

        if (this.tableFilters) {
          params = {
            ...params,
            ...this.tableFilters,
          };
        }

        if (this.tableSorter) {
          params = {
            ...params,
            sorter: `${this.tableSorter.field},${this.tableSorter.order}`,
          };
        }

        /**
         * 单选的表单项的参数传递
         */
        if (singleFields) {
          singleFields.forEach(singleField => {
            if (params[singleField]) {
              if (params[singleField].length > 0) {
                const name = singleFieldPrefix
                  ? `${singleFieldPrefix}${singleField
                      .slice(0, 1)
                      .toUpperCase()}${singleField.slice(1)}`
                  : singleField;
                params[name] = params[singleField][0];
              }
              delete params[singleField];
            }
          });
        }

        /**
         * 时间 range 参数传递
         */
        if (timeRangeFields) {
          timeRangeFields.forEach(timeRangeField => {
            if (params[timeRangeField]) {
              const [start, end] = params[timeRangeField];
              params[`${timeRangeField}Start`] = start;
              params[`${timeRangeField}End`] = end;
              delete params[timeRangeField];
            }
          });
        }

        console.log(params);

        // 将所有的参数放置到 url 中
        Router.replace({
          pathname,
          query: {
            ...params,
          },
        });

        dispatch({
          type: listRequestType,
          payload: {
            ...params,
          },
        });
      };

      public render() {
        return (
          <WrappedComponent
            {...this.props}
            handleTableChange={this.handleTableChange}
            handleSubmit={this.handleSubmit}
            handleSearch={this.handleSearch}
            urlQueryParams={this.urlQueryParams}
          />
        );
      }
    };
    return component as typeof WrappedComponent;
  };
}
