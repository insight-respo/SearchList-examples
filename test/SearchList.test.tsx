import { Component } from 'react';
import * as React from 'react';
import * as TestRenderer from 'react-test-renderer';
import * as _ from 'lodash'
import * as enzyme from 'enzyme';
import { Location } from 'history';
import SearchList from '@/SearchList';
import moment from 'moment';

/**
 * SearchList 组件依赖:
 * * dispatch
 * * location
 * * form antd 组件库的 form props, 可以 mock
 * * umi/router 组件允许使用 DO_NOT_USE_THIS_OPTIONS_ROUTER 进行覆盖
 */

function connect(dispatch: Function, location: Location, otherProps: object = {}) {
  return WrappedComponent => {
    return class extends Component {
      static displayName: string = `connect(${WrappedComponent.displayName || WrappedComponent.name})`
      public render() {
        return <WrappedComponent dispatch={dispatch} location={location} {...otherProps} />
      }
    } as typeof WrappedComponent
  }
}

const OriginComponent = function (props) {
    return <div>
      <div className="searchButton" onClick={props.handleSubmit}>form 表单搜索</div>
      <div onClick={props.handleTableChange}>表格</div>
    </div>
  }

test('should be render without error', () => {
  const dispatch = jest.fn(() => {})
  const location = {
    pathname: '',
    search: '',
    state: '',
    hash: '',
    query: {},
  }

  const compose = _.flowRight(
    connect(dispatch, location),
    SearchList({
      listRequestType: 'testDispatch/getList',
    })
  )

  const Component = compose(OriginComponent)

  const testRenderer = TestRenderer.create(<Component />)

  const tree = testRenderer.toJSON()

  expect(tree).toMatchSnapshot()
})

test('should: 从 location props 中正确的初始化参数(无需特殊处理的 url 参数)', () => {
  const mockQuery = {
    drugName: '阿司匹林',
    companyName: '阿里',
  }
  const dispatch = jest.fn(() => {})
  const location = {
    pathname: '',
    search: '',
    state: '',
    hash: '',
    query: {
      ...mockQuery
    },
  }

  const compose = _.flowRight(
    connect(dispatch, location),
    SearchList({
      listRequestType: 'testDispatch/getList',
    })
  )

  const Component = compose(OriginComponent)

  const testRenderer = TestRenderer.create(<Component />)
  const testInstance = testRenderer.root;

  expect(testInstance.findByType(OriginComponent).props.urlQueryParams).toEqual(mockQuery);
})

test('should: 从 location props 中正确的初始化参数, 含有特殊的 singleFields 参数', () => {
  const mockQuery = {
    isAssocStdElement: -1,
    isAssocStdIndication: 1,
  }

  const dispatch = jest.fn(() => {})
  const location = {
    pathname: '',
    search: '',
    state: '',
    hash: '',
    query: {
      ...mockQuery
    },
  }

  const compose = _.flowRight(
    connect(dispatch, location),
    SearchList({
      listRequestType: 'testDispatch/getList',
      singleFields: ['stdElement', 'stdIndication'],
      singleFieldPrefix: 'isAssoc',
    })
  )

  const Component = compose(OriginComponent)

  const testRenderer = TestRenderer.create(<Component />)
  const testInstance = testRenderer.root;

  expect(testInstance.findByType(OriginComponent).props.urlQueryParams).toEqual(mockQuery);
})

test('should: 从 location props 中正确的初始化参数, 含有特殊的 timeRangeFields 参数', () => {
  const mockQuery = {
    updateTimeStart: 3443,
    updateTimeEnd: 4324,
  }

  const dispatch = jest.fn(() => {})
  const location = {
    pathname: '',
    search: '',
    state: '',
    hash: '',
    query: {
      ...mockQuery
    },
  }

  const compose = _.flowRight(
    connect(dispatch, location),
    SearchList({
      listRequestType: 'testDispatch/getList',
      timeRangeFields: ['updateTime'],
    })
  )

  const Component = compose(OriginComponent)

  const testRenderer = TestRenderer.create(<Component />)
  const testInstance = testRenderer.root;

  expect(testInstance.findByType(OriginComponent).props.urlQueryParams).toEqual(mockQuery);
})

describe('should: 在点击搜索按钮的时候调用正确的函数和传入正确的参数', () => {
  test('should: 只入参 singleFields, singleFieldPrefix. 正确的调用 dispatch, validateFields, Router.replace.', () => {
    const mockQuery = {
      id: '123456',
    }
    const dispatch = jest.fn(() => {})
    const validateFields = jest.fn((callback) => {
      callback(null, {
        id: '123456',
      })
    })
    const routerReplace = jest.fn(() => {})
    const location = {
      pathname: '/trending',
      search: '',
      state: '',
      hash: '',
      query: {
        ...mockQuery
      },
    }
    const Router = {
      replace: routerReplace,
      push: () => {},
      go: () => {},
      goBack: () =>{},
    }
  
    const compose = _.flowRight(
      connect(dispatch, location, {
        form: {
          validateFields
        }
      }),
      SearchList({
        listRequestType: 'testDispatch/getList',
        singleFieldPrefix: 'isAssoc',
        singleFields: ['stdElement'],
        DO_NOT_USE_THIS_OPTIONS_ROUTER: Router,
      })
    )
  
    const Component = compose(OriginComponent)
  
    const testRenderer = enzyme.mount(<Component />);
  
    const originComponent = testRenderer.find(OriginComponent)
    originComponent.find('.searchButton').simulate('click')
  
    expect(validateFields).toBeCalled()
    expect(routerReplace).toBeCalledWith({
      pathname: '/trending',
      query: {
        p: 1,
        pageSize: 15,
        id: '123456',
      }
    })
    expect(dispatch).toBeCalledWith({
      type: 'testDispatch/getList',
      payload: {
        p: 1,
        pageSize: 15,
        id: '123456',
      },
    })
  })

  test('should: 只入参 timeRangeFields. 正确的调用 dispatch, validateFields, Router.replace', () => {
    const mockQuery = {
      id: '123456',
    }
    const dispatch = jest.fn(() => {});
    const validateFields = jest.fn((callback) => {
      callback(null, {
        updateTime: [moment(1560407706310), moment(1560407857997)],
      })
    })
    const routerReplace = jest.fn(() => {})
    const location = {
      pathname: '/trending',
      search: '',
      state: '',
      hash: '',
      query: {
        ...mockQuery
      },
    }
    const Router = {
      replace: routerReplace,
      push: () => {},
      go: () => {},
      goBack: () =>{},
    }

    const compose = _.flowRight(
      connect(dispatch, location, {
        form: {
          validateFields
        }
      }),
      SearchList({
        listRequestType: 'testDispatch/getList',
        timeRangeFields: ['updateTime', 'createTime'],
        DO_NOT_USE_THIS_OPTIONS_ROUTER: Router,
      })
    )
  
    const Component = compose(OriginComponent)
  
    const testRenderer = enzyme.mount(<Component />)

    const originComponent = testRenderer.find(OriginComponent)
    originComponent.find('.searchButton').simulate('click')

    expect(validateFields).toBeCalled()
    expect(routerReplace).toBeCalledWith({
      pathname: '/trending',
      query: {
        p: 1,
        pageSize: 15,
        id: '123456',
        updateTimeStart: 1560407706310,
        updateTimeEnd: 1560407857997,
      }
    })
    expect(dispatch).toBeCalledWith({
      type: 'testDispatch/getList',
      payload: {
        p: 1,
        pageSize: 15,
        id: '123456',
        updateTimeStart: 1560407706310,
        updateTimeEnd: 1560407857997,
      }
    })
  })
})

describe('should: 在表格变化的时候调用合适的函数和传递响应的参数', () => {
  test('should: 点击表格中单选项的 reset 按钮来重置筛选项', () => {
    const OriginComponent = function (props) {
      return (
        <div onClick={() => {
          props.handleTableChange({
            current: 1,
            pageSize: 15,
          }, {
            // ant design 中, 传递空数组来重置
            stdElement: [],
          })
        }} className="resetStdElement">重置 stdElement 筛选项的按钮</div>
      )
    }
    const mockQuery = {
      isAssocStdElement: 1,
    }
    const dispatch = jest.fn(() => {})
    const validateFields = jest.fn((callback) => { callback(null, {}) })
    const routerReplace = jest.fn(() => {})
    const location = {
      pathname: '/',
      search: '',
      state: '',
      hash: '',
      query: {
        ...mockQuery
      },
    }
    const Router = {
      replace: routerReplace,
      push: () => {},
      go: () => {},
      goBack: () =>{},
    }

    const compose = _.flowRight(
      connect(dispatch, location, {
        form: {
          validateFields
        }
      }),
      SearchList({
        listRequestType: 'testDispatch/getList',
        singleFieldPrefix: 'isAssoc',
        singleFields: ['stdElement'],
        DO_NOT_USE_THIS_OPTIONS_ROUTER: Router,
      })
    )
  
    const Component = compose(OriginComponent)
  
    const testRenderer = enzyme.mount(<Component />)

    const originComponent = testRenderer.find(OriginComponent)
    expect(originComponent.prop('urlQueryParams')).toEqual({
      isAssocStdElement: 1,
    })
    
    const resetBtn = originComponent.find('.resetStdElement')

    resetBtn.simulate('click')

    expect(dispatch).toBeCalledWith({
      type: 'testDispatch/getList',
      payload: {
        p: 1,
        pageSize: 15,
      },
    })

    expect(routerReplace).toBeCalledWith({
      pathname: '/',
      query: {
        p: 1,
        pageSize: 15,
      }
    })
  })
})