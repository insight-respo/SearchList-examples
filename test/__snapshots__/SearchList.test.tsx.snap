// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should be render without error 1`] = `
<div>
  <div
    className="searchButton"
    onClick={[Function]}
  >
    form 表单搜索
  </div>
  <div
    onClick={[Function]}
  >
    表格
  </div>
</div>
`;
